File: crates/fc-platform/src/application_openapi_spec/operations/events.rs
//! OpenAPI spec domain events.

use serde::{Deserialize, Serialize};

use crate::impl_domain_event;
use crate::usecase::domain_event::EventMetadata;
use crate::usecase::ExecutionContext;
use crate::TsidGenerator;

/// Emitted when an application syncs a new OpenAPI document, whether the
/// content was new (versionDelta=true) or byte-identical to the prior CURRENT
/// (versionDelta=false). The audit log keeps both cases for completeness.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ApplicationOpenApiSpecSynced {
    #[serde(flatten)]
    pub metadata: EventMetadata,

    pub application_id: String,
    pub application_code: String,
    pub spec_id: String,
    pub version: String,
    pub spec_hash: String,
    /// Some when a prior CURRENT was archived in this sync.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub archived_prior_version: Option<String>,
    /// True if the diff includes removed paths/schemas/verbs.
    pub has_breaking: bool,
    /// True if the incoming spec was byte-identical to the existing CURRENT
    /// (no new row was inserted).
    pub unchanged: bool,
}

impl_domain_event!(ApplicationOpenApiSpecSynced);

impl ApplicationOpenApiSpecSynced {
    const EVENT_TYPE: &'static str = "platform:developer:application-openapi:synced";
    const SPEC_VERSION: &'static str = "1.0";
    const SOURCE: &'static str = "platform:developer";

    #[allow(clippy::too_many_arguments)]
    pub fn new(
        ctx: &ExecutionContext,
        application_id: &str,
        application_code: &str,
        spec_id: &str,
        version: &str,
        spec_hash: &str,
        archived_prior_version: Option<String>,
        has_breaking: bool,
        unchanged: bool,
    ) -> Self {
        let event_id = TsidGenerator::generate_untyped();
        let subject = format!("platform.application-openapi.{}", spec_id);
        let message_group = format!("platform:application-openapi:{}", application_id);

        Self {
            metadata: EventMetadata::new(
                event_id,
                Self::EVENT_TYPE,
                Self::SPEC_VERSION,
                Self::SOURCE,
                subject,
                message_group,
                ctx.execution_id.clone(),
                ctx.correlation_id.clone(),
                ctx.causation_id.clone(),
                ctx.principal_id.clone(),
            ),
            application_id: application_id.to_string(),
            application_code: application_code.to_string(),
            spec_id: spec_id.to_string(),
            version: version.to_string(),
            spec_hash: spec_hash.to_string(),
            archived_prior_version,
            has_breaking,
            unchanged,
        }
    }
}

File: internal/platform/openapispecs/operations/events.go
// Package operations holds the OpenAPI-spec use cases. Each file is
// one verb; the shape mirrors the other platform operations packages.
package operations

import (
	"encoding/json"
	"time"

	"github.com/flowcatalyst/flowcatalyst-go/pkg/fcsdk/usecase"
)

const (
	SpecSyncedType = "platform:developer:application-openapi:synced"
	Source         = "platform:developer"
)

// ApplicationOpenApiSpecSynced is emitted after a sync — whether the
// document was new (Unchanged=false) or byte-identical to the prior
// CURRENT (Unchanged=true). The audit log keeps both for completeness.
type ApplicationOpenApiSpecSynced struct {
	Metadata             usecase.EventMetadata
	ApplicationID        string
	ApplicationCode      string
	SpecID               string
	Version              string
	SpecHash             string
	ArchivedPriorVersion *string
	HasBreaking          bool
	Unchanged            bool
}

func subjectFor(specID string) string { return "platform.application-openapi." + specID }
func groupFor(appID string) string    { return "platform:application-openapi:" + appID }

func (e ApplicationOpenApiSpecSynced) EventID() string       { return e.Metadata.EventID }
func (e ApplicationOpenApiSpecSynced) EventType() string     { return SpecSyncedType }
func (e ApplicationOpenApiSpecSynced) SpecVersion() string   { return "1.0" }
func (e ApplicationOpenApiSpecSynced) Source() string        { return Source }
func (e ApplicationOpenApiSpecSynced) Subject() string       { return subjectFor(e.SpecID) }
func (e ApplicationOpenApiSpecSynced) Time() time.Time       { return e.Metadata.OccurredAt }
func (e ApplicationOpenApiSpecSynced) PrincipalID() string   { return e.Metadata.PrincipalID }
func (e ApplicationOpenApiSpecSynced) CorrelationID() string { return e.Metadata.CorrelationID }
func (e ApplicationOpenApiSpecSynced) CausationID() string   { return e.Metadata.CausationID }
func (e ApplicationOpenApiSpecSynced) ExecutionID() string   { return e.Metadata.ExecutionID }
func (e ApplicationOpenApiSpecSynced) MessageGroup() string  { return groupFor(e.ApplicationID) }
func (e ApplicationOpenApiSpecSynced) ToDataJSON() ([]byte, error) {
	return json.Marshal(struct {
		ApplicationID        string  `json:"applicationId"`
		ApplicationCode      string  `json:"applicationCode"`
		SpecID               string  `json:"specId"`
		Version              string  `json:"version"`
		SpecHash             string  `json:"specHash"`
		ArchivedPriorVersion *string `json:"archivedPriorVersion,omitempty"`
		HasBreaking          bool    `json:"hasBreaking"`
		Unchanged            bool    `json:"unchanged"`
	}{
		e.ApplicationID, e.ApplicationCode, e.SpecID, e.Version, e.SpecHash,
		e.ArchivedPriorVersion, e.HasBreaking, e.Unchanged,
	})
}
